### PR TITLE
feat: limit channels to reduce MQ's memory footprint

### PR DESCRIPTION
**Pull request by heycalmdown**
Rationales

* The number of channels are not related to the performance to communicate with RabbitMQ
* But it affects a reliability when the channel has closed unexpectedly
* The channels consume bigger memory of MQ nodes than we expected

### CODE REVIEW

File: src/spec/amqp-channel-pool.spec.ts [hunk 1]
```diff
@@ -1,42 +1,51 @@
...
-  });
+  it('can use channel disposer', spec(async () => {
+    expect((amqpChannelPool as any).idleChannels.length).toEqual(0);
+    await amqpChannelPool.usingChannel(async channel => {
+      const xName = `spec.temp.${+new Date()}`;
+      await channel.assertExchange(xName, 'fanout', {autoDelete: true});
+      await channel.deleteExchange(xName);
+    });
+    expect((amqpChannelPool as any).idleChannels.length).toEqual(1);
+  }));
```
Comment (UniverseHan) [inline src/spec/amqp-channel-pool.spec.ts]: I can't understand what this test do. Should usingChange add a channel or arrow function do something? Is arrow function disposer?

**Comment by UniverseHan** [comment]
Is there test case for testing limiting channel pool?

File: src/spec/amqp-channel-pool.spec.ts (hunk 1 above)
Comment (heycalmdown) [reply src/spec/amqp-channel-pool.spec.ts]: `usingChannel` is using `AmqpChannelPoolService.getChannelDisposer()` .

File: src/services/amqp-channel-pool-service.ts [hunk 2]
```diff
@@ -102,12 +93,10 @@ export class AmqpChannelPoolService {
         if (err.stack) {
           logger.debug(err.stack);
         }
+        _.remove(this.idleChannels, channel);
```
[the remark this answers is not in the document]
Comment (heycalmdown) [reply src/services/amqp-channel-pool-service.ts]: The limitation is not about the AMQP, just for Java implementation.

[the remark this answers is not in the document]
Comment (resoliwan) [reply src/services/amqp-channel-pool-service.ts]: Ah! Thanks.
Channel is a logical connection between a server and a mq.
Why are we share same channel for different tasks in same time? 
I think it adds more complicate when we follow the logic and it also
makes hard to debug.

[the remark this answers is not in the document]
Comment (heycalmdown) [reply src/services/amqp-channel-pool-service.ts]: The reason is `many channels affect MQ node's memory`.

[the remark this answers is not in the document]
Comment (resoliwan) [reply src/services/amqp-channel-pool-service.ts]: approved

**Pull request merged**